Fixed #8866: simplification of log function.

alkalinin opened the pull request.

alkalinin committed:
Fixed #8866: log function with evaluate=False

Fixed simplify() and expand_log() operations of a log() function with not
default base and evaluate=False flag.

aktech commented:
I think this doesn't fixes the problem in #8866 :
as the following error persists, even after the above code 

```python
In [3]: simplify(log(x, 10, evaluate=False))
Out[3]: log(x)

In [4]: simplify(log(x, 10))
Out[4]: log(x)/log(10)
```
and please add tests.

smichr commented on sympy/functions/elementary/exponential.py:
```diff
@@ -596,7 +596,8 @@ def _eval_expand_log(self, deep=True, **hints):
     def _eval_simplify(self, ratio, measure):
         from sympy.simplify.simplify import expand_log, simplify
-        expr = self.func(simplify(self.args[0], ratio=ratio, measure=measure))
+        expr = self.func(simplify(self.args[0], ratio=ratio, measure=measure),
```
`base` is known to eval but I think is ignored at instantiation; I'm not sure why you don't get an error (or maybe you will when tests run).  How about

```
expr = self.func(*[simplify(i, ratio=ratio, measure=measure)
    for i in self.args])
```

alec-kalinin replied:
I've checked this solution, it works! I think this is the best way to solve the problem.

What should I do with this pull request? Also I want to add  the following test to the
sympy/sympy/simplify/tests/test_simplify.py
```python
def test_issue_8866():
    assert simplify(log(x, 10, evaluate=False)) == simplify(log(x, 10))
```

alec-kalinin commented:
I've checked this solution, it works! I think this is the best way to solve the problem.

What should I do with this pull request? Also I want to add  the following test to the
sympy/sympy/simplify/tests/test_simplify.py
```python
def test_issue_8866():
    assert simplify(log(x, 10, evaluate=False)) == simplify(log(x, 10))
```

aktech commented:
Simply make changes (add tests also) in your local branch `basemaster` and push it to the same branch, your pull request will be updated with new changes.
You can also squash commits using [git rebase](https://help.github.com/articles/about-git-rebase/).

alec-kalinin commented:
Sorry, I did more tests and found that the solution
```python
expr = self.func(*[simplify(i, ratio=ratio, measure=measure) for i in self.args])
```
still not working:
```python
>>> simplify(log(x, 10))
log(x)/log(10)
>>> simplify(log(x, 10, evaluate=False))
log(x, 10)
>>> expand_log(simplify(log(x, 10, evaluate=False)))
log(x)
```

The problem is still in _eval_simplify function in return statement:
```python
return min([expr, self], key=measure)
```
The expr is correct: log(x)/log(10), but self = log(x, 10) and log(x, 10) < log(x)/log(10) and function return log(x, 10).

Now I am thinking what is the best way to fix this...

smichr commented:
How about

```diff
diff --git a/sympy/functions/elementary/exponential.py b/sympy/functions/elementary/exponential.py
index 1b1c03f..59ff51c 100644
--- a/sympy/functions/elementary/exponential.py
+++ b/sympy/functions/elementary/exponential.py
@@ -554,49 +554,52 @@ def taylor_term(n, x, *previous_terms):  # of log(1+x)
     def _eval_expand_log(self, deep=True, **hints):
         from sympy import unpolarify
         from sympy.concrete import Sum, Product
+        func = lambda x: self.func(x, self.args[1], evaluate=False) if \
+            len(self.args) == 2 else self.func(x)
         force = hints.get('force', False)
         arg = self.args[0]
         if arg.is_Integer:
             # remove perfect powers
             p = perfect_power(int(arg))
             if p is not False:
-                return p[1]*self.func(p[0])
+                return p[1]*func(p[0])
         elif arg.is_Mul:
             expr = []
             nonpos = []
             for x in arg.args:
                 if force or x.is_positive or x.is_polar:
-                    a = self.func(x)
+                    a = func(x)
                     if isinstance(a, log):
-                        expr.append(self.func(x)._eval_expand_log(**hints))
+                        expr.append(func(x)._eval_expand_log(**hints))
                     else:
                         expr.append(a)
                 elif x.is_negative:
-                    a = self.func(-x)
+                    a = func(-x)
                     expr.append(a)
                     nonpos.append(S.NegativeOne)
                 else:
                     nonpos.append(x)
-            return Add(*expr) + log(Mul(*nonpos))
+            return Add(*expr) + func(Mul(*nonpos))
         elif arg.is_Pow or isinstance(arg, exp):
             if force or (arg.exp.is_real and arg.base.is_positive) or \
                     arg.base.is_polar:
                 b = arg.base
                 e = arg.exp
-                a = self.func(b)
+                a = func(b)
                 if isinstance(a, log):
                     return unpolarify(e) * a._eval_expand_log(**hints)
                 else:
                     return unpolarify(e) * a
         elif isinstance(arg, Product):
             if arg.function.is_positive:
-                return Sum(log(arg.function), *arg.limits)
+                return Sum(func(arg.function), *arg.limits)

-        return self.func(arg)
+        return func(arg)

     def _eval_simplify(self, ratio, measure):
         from sympy.simplify.simplify import expand_log, simplify
-        expr = self.func(simplify(self.args[0], ratio=ratio, measure=measure))
+        expr = self.func(*[simplify(i, ratio=ratio, measure=measure)
+            for i in self.args])
         expr = expand_log(expr, deep=True)
         return min([expr, self], key=measure)
```

I'm not sure about making func use evaluate=False...see what you think.

```python
>>> expand_log(log(x,10,evaluate=0))
log(x, 10)
>>> var('x y',positive=1)
(x, y)
>>> expand_log(log(x*y,10,evaluate=0))
log(x, 10) + log(y, 10) + log(1, 10)
```

alec-kalinin commented:
I am not very sure that making function use evaluate=False is a good solution. My use case is following:
1. I need evaluate=False only to print out the original equation. I need this for nice solution logs.
2. After that I prefer to evaluate and unfold all functions.
But may be keeping evaluate=False is a right way.

And what about an idea to evaluate all logarithm functions as a first step of transformations:
```python
    def _eval_simplify(self, ratio, measure):
        if (len(self.args) == 2):
            return self._eval_simplify(self.func(*[self.args]), 
                                       ratio=ratio, measure=measure)
        
        from sympy.simplify.simplify import expand_log, simplify
        expr = self.func(simplify(self.args[0], ratio=ratio, measure=measure))
        expr = expand_log(expr, deep=True)
        return min([expr, self], key=measure)
```

```python
    def _eval_expand_log(self, deep=True, **hints):
        if (len(self.args) == 2):
            from sympy import expand_log
            return expand_log(self.func(*[self.args]),     # I know this is not working, it is just an idea
                              deep=True, **hints)
    ...
```

smichr commented:
The general behavior of evaluated=False is that the object becomes evaluated once it is "touched" (e.g. multiplied by something, etc...). So perhaps *not* using unevaluated=False in expand_log is the better way to go; and I like the idea of evaluating in simplify.

smichr commented:
Perhaps the _eval_expand_log can just start out as:

```
if len(self.args) == 2
    return expand_log(self.func(*self.args))
```

alec-kalinin commented:
I agree with you. I've committed new fix. Now my tests looks good:
```python
>>> simplify(log(x, 10, evaluate=False))
log(x)/log(10)
>>> expand_log(log(x, 10, evaluate=False))
log(x)/log(10)
>>> expand(log(x*(y + z), 10, evaluate=False))
log(x*y + x*z)/log(10)
```

smichr commented:
One final thought: perhaps a test that affirms that the base has been affected by expansion or simplification would be good. Maybe this:

```python
>>> var('x', positive=True)
x
>>> l = log(exp(x), exp(10), evaluate=False)
>>> b = log(exp(x), exp(5), evaluate=False)
>>> expand_log(log(l, b))
(log(x) - log(10))/(log(x) - log(5))
>>> simplify(log(l, b))
log((x/10)**(1/log(x/5)))
```

alec-kalinin commented:
Is it OK to compare in test just evaluated logarithms versus "evaluate=False" logarithms including test of transformations of the base? The code is:
```Python
def test_issue_8866():
    x = Symbol('x')
    assert simplify(log(x, 10, evaluate=False)) == simplify(log(x, 10))
    assert expand_log(log(x, 10, evaluate=False)) == expand_log(log(x, 10))

    y = Symbol('y', positive=True)
    l1 = log(exp(y), exp(10))
    b1 = log(exp(y), exp(5))
    l2 = log(exp(y), exp(10), evaluate=False)
    b2 = log(exp(y), exp(5), evaluate=False)
    assert simplify(log(l1, b1)) == simplify(log(l2, b2))
    assert expand_log(log(l1, b1)) == expand_log(log(l2, b2))
```

smichr commented:
That test looks good.

alkalinin committed:
Fixed #8866: log function with evaluate=False

Added test that confirm that the base also is affected by simplifications
and expansions.

alec-kalinin commented:
Test has been committed.

smichr commented:
OK, it's in. Thanks!
